### PR TITLE
Move pattern match for foreach before bulky if-else in `save_variables`

### PR DESCRIPTION
**Pull request by crcrpar**
One caveat could be that the first if branch doesn't seem to use `arg.expr` at all.

fixes https://github.com/pytorch/pytorch/pull/96405#discussion_r1175669480.

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -1384,17 +1394,16 @@ def save_variables(
                     is_inplace_view = f"{var}.is_view()"
                     expr = f"SavedVariable({var}, {str(is_output).lower()}, {is_inplace_view})"
                 else:
-                    var_ = var + ("[i]" if is_foreacharg_list_type else "")
                     expr = f"SavedVariable({var}, {str(is_output).lower()})"
-                remove_underscore_before_replace = True and "self" not in var
+                if foreacharg is not None and "original_selfs" not in expr:
```
Comment (soulitzer) [inline tools/autograd/gen_variable_type.py:1398]: We could still do this in the 'else' right?

**Review by soulitzer — APPROVED** [review]
Thanks! Just had a comment about where to put the expr replacement in the Tensor case.

**Comment by crcrpar** [comment]
https://github.com/pytorch/pytorch/actions/runs/4864414689/jobs/8673892924#step:13:3829 🤔 

```
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/quantization/mobilenetv2.py", line 152, in mobilenet_v2
    model.load_state_dict(weights.get_state_dict(progress=progress))
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/_api.py", line 89, in get_state_dict
    return load_state_dict_from_url(self.url, progress=progress)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/hub.py", line 750, in load_state_dict_from_url
    download_url_to_file(url, cached_file, hash_prefix, progress=progress)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/hub.py", line 612, in download_url_to_file
    u = urlopen(req)
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 519, in open
    response = self._open(req, data)
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 1391, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/opt/conda/envs/py_3.10/lib/python3.10/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno -2] Name or service not known>
```

**Pull request closed without merging**